### ISSUE TITLE
GPS altitude filtering/smoothing (feature)

### ISSUE THREAD

**Issue by luszczyszenko** [bailuk/AAT#101]
I know that mobile phone's GPS has limited accuracy. And if I put the phone in the pocket there's only worse. 
But... Is there a possibility to "smooth" logged GPS altitude data? Maybe moving average of a few last samples? Some kind of low pass filter?

Current situation is that "climb/descend" parameter counts every single meter up/down but these are not real values.

**Comment by bailuk** [comment]
> Current situation is that "climb/descend" parameter counts every single meter up/down but these are not real values.

This is not the case: AAT uses [AverageAltitude.java](https://github.com/bailuk/AAT/blob/master/app/src/main/java/ch/bailu/aat/gpx/attributes/AverageAltitude.java) to average altitude before calculating deltas. Its my impression that total descend and ascend give quite realistic values. But personally I'm not interested in super precise values.

These are the limits used to average GPS samples:
``` Java
private static final int MIN_SAMPLES = 5;
private static final int MAX_SAMPLES = 50;
private static final float MIN_DISTANCE = 50f;
```

To get more precise altitude you can adjust GPS value (in preferences) with a fixed correction value or enable the phones barometer sensor and then manually set altitude whenever you come to post sign.

BTW if you have _Adjust altitude_ or _barometer sensor_ enabled you can just tab _altitude_ in the cockpit view to adjust altitude.

**Comment by luszczyszenko** [comment]
Yes, in you got me. I did not checked if any averaging of altitude is done. 

> AAT uses AverageAltitude.java to average altitude before calculating deltas.

But why it gives so poor result? Is my GPS so inaccurate? 


> To get more precise altitude you can adjust GPS value (in preferences) with a fixed correction value or enable the phones barometer sensor and then manually set altitude whenever you come to post sign.
> 
That's not solution. It makes only an offset.
I made a test loop:

![tmp-cam-6213303266414692049](https://user-images.githubusercontent.com/69072994/89721979-cc808600-d9e3-11ea-8598-b159939d9fce.jpg)
![tmp-cam-2116269647347708845](https://user-images.githubusercontent.com/69072994/89725224-804a3b80-da0d-11ea-9149-877cb26ff6bb.jpg)
AAT shows ~60m of climb / downhill. It's not real values.
It should be about 20m.
So I made similar loop track with brouter.de to show you the real values (or close to real) of altitude changes.
![tmp-cam-1555811194888750696](https://user-images.githubusercontent.com/69072994/89725274-fe0e4700-da0d-11ea-9027-7a1e93e9e794.jpg)
![tmp-cam-1156041096535002988](https://user-images.githubusercontent.com/69072994/89725226-86401c80-da0d-11ea-82dd-64d5ec63c37e.jpg)

Because start point is the end point there should be identical values of climb and downhill. But they're different. Why? There is no influence of GPS accuracy.
![merged](https://user-images.githubusercontent.com/69072994/89725612-95c16480-da11-11ea-8ebb-96508e6cf0bd.png)

**Comment by bailuk** [comment]
The BRouter variant is much smaller. It has only 87 samples while your track has 246. As one can see on the graph BRouter leaves away many smaller climbs. To me 60 m look realistic, if you add the smaller climbs. 

> Because start point is the end point there should be identical values of climb and downhill. But they're different. Why? There is no influence of GPS accuracy.

You're right, climb and downhill should be the same. The difference comes probably from the low resolution of the route in combination with averaging.

Some notes to clarify: 
- The altitude AAT records and saves to the GPX track are identical with altitude received from GPS.
- So what you see on the graph is identical with what was received from GPS.
- Only climb and downhill are averaged and the averaging is done when loading a GPX track.

**Comment by luszczyszenko** [comment]
>To me 60 m look realistic, if you add the smaller climbs.

I cannot agree. These are plain streets, not off-road. There are a lot less smaller climbs (and downhills) bythan GPS claims.
Next example. It is downhill street, so there is no climb.
![tmp-cam-7956243819561247484](https://user-images.githubusercontent.com/69072994/89758639-92e57300-dae8-11ea-935c-61c8052f4da7.jpg)
![tmp-cam-5962130822654951576](https://user-images.githubusercontent.com/69072994/89758654-9973ea80-dae8-11ea-93a0-51966681601f.jpg)


The main problem is in GPS accuracy, so it seems to me that nothing can be done with it. Ergo: climb and descend are unusable.
